### ISSUE TITLE
[5.3] Type narrowing regression with property comparison to boolean false

### ISSUE THREAD

**Issue by meyfa** [microsoft/TypeScript#56482]
### 🔎 Search Terms

narrowing, strict equals, property access, boolean false, undefined

### 🕗 Version & Regression Information

- This changed between versions 5.2 and 5.3
- This also happens in the nightly

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.1-rc#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIBGcUAXEZgDYRwi4C+uuC6IAzmMgOYRgZYBeZAAoAlOX7IAPsgCuIACYQYoCAuQCAfDnzIovWVBDIAsnDAALAHRRaC9AFsxybQAYrAVmQB+HERLk8JRsKPTI5PJKKiBqDEws7JwwmBrcvPxiuMAwIsno3lbEUBoCQkEhojoECWzo1FaU6FzCeaJxuAD0HcgAAmBsALQQAB4ADhAIYENQUOjFA8h5yMBsyKPobGzAhJQAnnKKyqoKzKy19Y3NeYUkbUA

### 💻 Code

```ts
interface Foo {
  bar: boolean
}

const getFoo = (): Foo | undefined => {
  return Math.random() > 0.5 ? { bar: false } : undefined
}

const foo = getFoo()
if (foo?.bar === false) {
  console.log(foo)
}

// @ts-expect-error - foo is possibly undefined
console.log(foo.bar)
```

### 🙁 Actual behavior

In line 13 and following, `foo` is incorrectly narrowed to just `Foo` (whereas it was previously `Foo | undefined`), although we learned nothing about it from the preceding `if` as that did not influence control flow.

```
Unused '@ts-expect-error' directive.(2578)
```

### 🙂 Expected behavior

TS should report an error when accessing `foo.bar` as `foo` can potentially be undefined.

### Additional information about the issue

I originally discovered this issue in JSX with a snippet similar to:

```tsx
function Component () {
  const foo = getFoo()
  return (
    <div>
      {foo?.bar === false && 'foo'}
      {foo.bar ? 'true' : 'false'}
    </div>
  )
}
```

There as well, the access `foo.bar` is allowed due to the preceding check, although they are completely separate.

Note that changing it to `=== true` or removing the checks entirely fixes the problem.

**Comment by Andarist** [comment]
It looks very much related to my change: https://github.com/microsoft/TypeScript/pull/53714 . I'll investigate this soon. Thanks for the report!

**Comment by akwodkiewicz** [comment]
Here's a different reproduction resembling the one we have in our project, maybe you find it helpful:

```
export interface Options {
    a?: boolean;
    b?: boolean;
}

function foobar(options?: Options) {
    // @ts-expect-error 'options' is possibly 'undefined'
    if (options?.a === false || options.b) {
    }
}
```

https://www.typescriptlang.org/play?ts=5.3.2#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgeTBgQiQGc4BvAKDjrkwH4AuOAIwggBthMkBuWvTYt2nHn0EBfatXQBXJNmKk46Tm0xQAFBCIlyowivIBKKkLoB6K3AACMMgFpQYYMpdQo0OAHI9JmS+iBSQZGQIbFwAnn6KACbA6MjA8b6WiOhwuvqkZIwAdJhwALxlaphcZHgAPjVwAQZkBWzmNPRwMlJAA

**Comment by jakebailey** [comment]
This issue has a lot of thumbs up; out of curiosity, are people finding this naturally? Or is there some Twitter thread that's directing people here?

The original PR didn't fail any of the extended test suites, so it's somewhat surprising to me to have a regression like this.

**Comment by Andarist** [comment]
My guess is that it was posted on a company’s Slack channel or smth and a couple of devs from that channel +1ed this 😉

**Comment by TrippTrapp84** [comment]
Looks like this issue still exists for some specific cases
https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEkoGdnwILwN4CgCQAksugBQCUAXPAC4AWAlmoxgNw4C+OOoSciA9gDtk1eAAco1aiBiCqmAD7wAroNAAzeoJDA261WGr0h8KGLEQAngAVJ02eWz566+CQlSZggPwA6IqRk8AC8ofDqUBDIIEG4eHge9oJs8QD0qfAAet74eHDUyrIpnAl2XinpWTnsQA

**Comment by jakebailey** [comment]
Please file a new issue
